File: jianzhi_offer(剑指offer)/序列化二叉树.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstring>
#include <queue>
#include <sstream>
using namespace std;

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};

class Solution {
    public:
    void SerializeHelper(TreeNode *r, string& str){
        str += ',';
        if(r == NULL) str += '$';
        else {
            //str += i2s(r->val);
            str += i2s_v2(r->val);
            SerializeHelper(r->left, str);
            SerializeHelper(r->right, str);
        }
        cout << str << endl;
    }
    char* Serialize(TreeNode *root) {
        string str;
        SerializeHelper(root, str);
        /*
        char* res = new char[str.size()+1];
        strcpy(res, str.c_str());
        ++res;  // escape one ','
        */

        //*
        char* res = new char[str.size()];
        char *dst = res;
        const char *src = (str.c_str())+1;  // escape one ','
        while ((*dst++=*src++)!='\0');
        //*/
        return res;
    }

    TreeNode* DeserializeHelper(TreeNode* p, char*& str) {
        if(*str == '\0') return NULL;
        if(*str == '$'){
            if(*(++str) == ','){
                ++str;  // 0, 1,2,$,$,$,$'\0'
            }
            return NULL;
        }
        p = new TreeNode(s2i(str));
        p->left = DeserializeHelper(p->left, str);
        p->right = DeserializeHelper(p->right, str);
        return p;
    }
    TreeNode* Deserialize(char *str) {
        return DeserializeHelper(NULL, str);
    }
private:
    string i2s(long long x){
        if(x == 0) return "0";
        string str;
        bool negative = false;
        if(x<0){
            negative = true;
            x = -x;
        }
        while(x != 0){
            str += char(x%10 + '0');
            x /= 10;
        } if(negative) str += '-';
        std::reverse(str.begin(), str.end());
        return str;
    }
    string i2s_v2(int x){
        stringstream ss;
        ss << x;
        string res;
        ss >> res;
        return res;
    }
    int s2i(char*& s){
        int res = 0;
        bool negative = false;
        if(*s == '-'){
            negative = true;
            ++s;
        } while(*s != ',' && *s != '\0'){
            res = res*10 + (*s++ - '0');
        } if(negative) res = -res;
        if(*s == ',') ++s;
        return res;
    }
};
TreeNode* constructTree(int& v, int x){
    if(v>x) return NULL;
    TreeNode* r = new TreeNode(v++);
    r->left = constructTree(v, x);
    r->right = constructTree(v, x);
    return r;
}
void printTreeNode(TreeNode* r){
    if(r){
        cout << r->val << endl;
        printTreeNode(r->left);
        printTreeNode(r->right);
    }
}
int main(){
    Solution solution = Solution();
    int cur_v = 0;
    //cout << solution.i2s(-1234) << endl;
    //cout << solution.s2i("-123") << endl;
    TreeNode* r = constructTree(cur_v, 100);
    printTreeNode(r);
    char* str_tree =  Solution().Serialize(r);
    cout << endl;
    cout << endl;
    cout << str_tree << endl;
    //cout << endl;
    //printTreeNode(solution.Deserialize(str_tree));

    return 0;
}

File: jianzhi_offer(剑指offer)/序列化二叉树.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.ArrayList;

class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;
    }
}

public class Solution {
    String Serialize(TreeNode root) {
        if(root == null) return "{}";
        ArrayList<String> arr = new ArrayList<String>();
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        q.offer(root);
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            if(node == null){
                arr.add("#");
            } else{
                arr.add(Integer.toString(node.val));
                q.offer(node.left);
                q.offer(node.right);    
            }
        }
        return "{"+String.join(",", arr)+"}";
    }
    TreeNode Deserialize(String str) {
        String [] nodes = str.replaceAll("(^\\{)|(}$)", "").split(",");
        if(nodes.length == 1 && nodes[0].equals("")) return null;
        TreeNode r = new TreeNode(Integer.parseInt(nodes[0]));
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        q.add(r);
        int idx = 1;
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            String s;
            for(int i = 0; i < 2; ++i){
                if(idx == nodes.length) break;
                s = nodes[idx++];
                if(s.equals("#")) continue;
                if(i == 0){
                    node.left = new TreeNode(Integer.parseInt(s));
                    q.add(node.left);
                } else{
                    node.right = new TreeNode(Integer.parseInt(s));
                    q.add(node.right);
                }
            }
        }
        return r;
    }
    public static void main(String[] args){
        String [] arr = {
                "{1,2,3,#,#,6,7}",
                "{1}",
                "{}"
        };
        Solution obj = new Solution();
        for(String str: arr){
            TreeNode node = obj.Deserialize(str);
            String res = obj.Serialize(node);
            System.out.println(res);
        }
    }
}